Temporal Nexus: Encyclopedia - Worker and Workflow pgs (#3055)
temporalio/documentation: docs/encyclopedia/workers.mdx

## What is a Task? {#task}

> A Task is the context that a Worker needs to progress with a specific [Workflow Execution](/workflows#workflow-execution), [Activity Execution](/activities#activity-execution), or a [Nexus Operation Task Execution](#nexus-operation-task-execution).

Reviewer: ```suggestion
A Task is the context that a Worker needs to progress with a specific [Workflow Execution](/workflows#workflow-execution), [Activity Execution](/activities#activity-execution), or a [Nexus Task Execution](#nexus-task-execution).
```

There are three types of Tasks:

- [Activity Task](#activity-task)
- [Workflow Task](#workflow-task)

> - [Nexus Operation Task](#nexus-operation-task)

Reviewer: ```suggestion
- [Nexus Task](#nexus-task)
```

- ActivityTaskTerminated
- ActivityTaskTimedOut

> ## What is a Nexus Operation Task? {#nexus-operation-task}

Reviewer: ```suggestion
## What is a Nexus Task? {#nexus-task}
```

A Nexus Operation Task provides the context needed to proceed with an Nexus Operation Task Execution.

> Nexus Operation Tasks largely represent the Nexus Operation Scheduled Event, containing the data needed to execute a Nexus Operation handler.

Reviewer: A Nexus Task is a representation of a single Nexus request to start or cancel a Nexus Operation. The task includes information such as the Service and Operation names and other information required to process the Nexus request.
The Temporal Worker triggers the registered Operation handler based on the task's information.

Author: Thanks!

> ## What is a Nexus Operation Task Execution? {#nexus-operation-task-execution}

Reviewer: ```suggestion
## What is a Nexus Task Execution? {#nexus-task-execution}
```

> An Nexus Operation Task Execution occurs when a Worker uses the context provided from the Nexus Operation Task and executes the Nexus Operation handler (also known as the Nexus Operation handler function).

Reviewer: ```suggestion
A Nexus Task Execution occurs when a Worker uses the context provided from the Nexus Operation Task and executes the Nexus Operation handler (also known as the Nexus Operation handler function).
```

> An Nexus Operation Task Execution occurs when a Worker uses the context provided from the Nexus Operation Task and executes the Nexus Operation handler (also known as the Nexus Operation handler function).

Reviewer 2: ```suggestion
A Nexus Task Execution occurs when a Worker uses the context provided from the Nexus Task and executes an action associated with a Nexus Operation, which commonly includes starting a Nexus Operation using it's Nexus Operation handler plus many additional actions that may be performed on a Nexus Operation.
```

> The NexusOperationScheduled Event corresponds to when the Temporal Service puts the Nexus Operation Task into the Task Queue.

Reviewer 3: ```suggestion
The `NexusOperationScheduled` Event corresponds to when the Temporal Service puts the Nexus Operation Task into the Task Queue.
```

Should these be backticked?

> The NexusOperationScheduled Event corresponds to when the Temporal Service puts the Nexus Operation Task into the Task Queue.

Reviewer: This is incorrect, please remove.

> The NexusOperationScheduled Event corresponds to when the Temporal Service puts the Nexus Operation Task into the Task Queue.

Reviewer 2: ```suggestion
The NexusOperationScheduled Event corresponds to when the Temporal Service puts the Nexus Task to start a Nexus Operation into the Task Queue.
```

Reviewer: The NexusOperationScheduled Event corresponds to when the Temporal Service records the Workflow's intent to schedule an operation.

> The NexusOperationStarted Event corresponds to when the Worker picks up the Nexus Operation Task from the Task Queue.

Reviewer: Also incorrect.

> The NexusOperationStarted Event corresponds to when the Worker picks up the Nexus Operation Task from the Task Queue.

Reviewer 2: ```suggestion
The NexusOperationStarted Event corresponds to when the Worker picks up the Nexus Task from the Task Queue, starts an asynchronous Nexus Operation, and returns an Operation ID to the caller indicating the asynchronous Nexus Operation has started.
```

> Either NexusOperationCompleted or one of the other Closed Nexus Operation Events corresponds to when the Worker has yielded back to the Temporal Service.

Reviewer: Also incorrect.

> Either NexusOperationCompleted or one of the other Closed Nexus Operation Events corresponds to when the Worker has yielded back to the Temporal Service.

Reviewer 2: ```suggestion
Either NexusOperationCompleted or one of the other Closed Nexus Operation Events corresponds to when the Nexus Operation has reached a final state due to successfully completing the operation or unsuccessfully completing the operation in the case of a failure, timeout, or cancelation.
```

> The API to schedule a Nexus Operation Execution provides an "effectively once" experience, even though there may be several Nexus Operation Task Executions that take place to successfully complete an Nexus Operation.

Reviewer: Incorrect.

> The API to schedule a Nexus Operation Execution provides an "effectively once" experience, even though there may be several Nexus Operation Task Executions that take place to successfully complete an Nexus Operation.

Reviewer 2: ```suggestion
The API to schedule a Nexus Operation Execution may provide an "effectively once" experience, depending on the implementation of the Nexus Operation handler (which should be idempotent), even though there may be several Nexus Task Executions that take place to successfully complete an Nexus Operation, for example in the case of retries.
```

Reviewer: How about this?

A Nexus Operation Execution appears to the caller Workflow as a single RPC, while under the hood the Temporal Service may issue several Nexus Tasks to attempt to start the Operation. Hence, a Nexus Operation Handler implementation should be idempotent. The WorkflowRunOperation provided by the SDK leverages Workflow ID based deduplication to ensures  idempotency and provide an "effectively once" experience.

> Once an Nexus Operation Task finishes execution, the Worker responds to the Temporal Service with a specific Event:

Reviewer: Inaccurate.

> Once an Nexus Operation Task finishes execution, the Worker responds to the Temporal Service with a specific Event:

Reviewer 2: ```suggestion
Once an Nexus Task finishes execution, the Worker responds to the Temporal Service with a specific Event:
```

> Once an Nexus Operation Task finishes execution, the Worker responds to the Temporal Service with a specific Event:

Reviewer: ```suggestion
A Nexus Task Execution completes when a Worker responds to the Temporal Service with either a RespondNexusTaskCompleted or RespondNexusTaskFailed call, or when the task times out. Temporal Service interprets the outcome and determines whether to retry the task or record the progress in a History Event:
```

> - NexusOperationCompleted

Reviewer 2: ```suggestion
- NexusTaskCompleted
```

Reviewer: This was correct @prasek

Reviewer: I'd also add NexusOperationStarted which denotes that the Handler responded with an Operation ID, indicating the Operation is asynchronous and will be completed using a Completion Callback.

> - NexusOperationFailed

Reviewer 2: ```suggestion
- NexusTaskFailed
```

Reviewer 2: Should we document [this level](https://github.com/temporalio/api/blob/d761393206a8d6d41878fe436a8690ee7a21783d/temporal/api/workflowservice/v1/request_response.proto#L1593) of the Nexus internals, since they are one level below what user's typically interact with?

Reviewer: This was also correct, there's no NexusTaskFailed event.

> - NexusOperationTimedOut

Reviewer 2: ```suggestion
```

Reviewer: TimedOut is somewhat relevant but doesn't happen due to a single task execution.

> - NexusOperationCanceled

Reviewer 2: ```suggestion
```

Reviewer: Could be resolved as Canceled.
